proposals/XXXX-extend-property-wrappers-to-functions-and-closures.md

```diff
@@ -0,0 +1,391 @@
+
+```swift
+@propertyWrapper
+struct Wrapper<Value> {
+    var wrappedValue: Value
+}
+
+struct Foo {
+    @Wrapper
+    var count = 0
+     
+    func increase() {
+        count += 1
+        // Great!
+    }
+}
+
+func foo(count: Wrapper<Int>) {
+    count.wrappedValue = ...
+    //   ^~~~~~~~~~~~ 
+    // Unfortunately, we can't
+    // use `@Wrapper` here.
+}
+```
+
+As seen in the above example, it quite akward and unintuitive that property wrappers cannot be applied to function parameters. This is only emphasized by the fact that property wrappers originally sought out to abstract away such accessor patterns.  As a result, elegant APIs are undermined by this limitation. Not only, is this limiting users by forcing them to rigidly follow API guidelines, which may not cover a specific use case, but it also limits API authors in what they can create. That is, API authors can't use property-wrapper types in closure parameters nor can code be seperated into functions that accept property wrapper syntax:
+
+```swift
+func fooInClosure(
+    _ block: (Wrapper<Int>) -> Void
+) { ... }
+
+fooInClosure { count in 
+    count.wrappedValue = 2
+    //    ^~~~~~~~~~~~ 
+    // Again, we have to 
+    // access count through
+    // `wrappedValue`.
+}
+```
+
+In fact, establishing custom behavior on closure parameters is really powerful. For example, if such a feature were supported, it could be used in conjunction with [Function Builders](https://github.com/apple/swift-evolution/blob/master/proposals/0289-function-builders.md) to expose data managed by a 'component' type. Moreover, property wrappers in escaping closures could be used to expose data available at the time the closure is executed in a simple and intuitive manner.
+
+## Proposed solution
+
+We propose to extend the contexts were application of property-wrapper types is allowed. Namely, application of such types will be allowed on function and closure parameters:
+
+```swift
+@propertyWrapper
+struct Wrapper<Value> {
+    var wrappedValue: Value
+    
+    var projectedValue: Self {
+        self
+    }
+}
+
+func foo(@Wrapper count: Int = 0) {
+    ...
+}
+
+func fooInClosure(
+    _ block: (Wrapper<Int>) -> Int
+) { ... }
+
+struct Foo {
+    @Wrapper
+    var count = 0
+    
+    func bar() {
+        foo(count: $count)
+        
+        fooInClosure { @Wrapper count in
```

Comment:
Would it be possible to support attaching it to the closure parameter at declaration site like you do for non-closure parameters? For example - `fooInClosure(@Wrapper block: (Wrapper<Int>) -> Int)`, because it would allow a user to use anonymous parameters and access count via `$0` for example.

Reply:
That wouldn't work if you have multiple property wrappers parameters in the function type. Listing the wrapper attributes for different parameters next to each other before the parameter also wouldn't mesh well with property wrapper composition.

I'm against inferring the property wrapper attribute on closure parameters for several reasons. I also think it's useful to not be forced into the property wrapper syntax. However, I think we should discuss design decisions on the forums

Reply:
Also, I would _really_ hate to have to refer to the backing wrapper and projected value with `_$0` and `$$0` 😆

Reply:
Oh, okay! It might be good to add an example involving anonymous parameters and mentioning that it would make accessing the backing and projected value a bit awkward :)

Reply:
✅